### ISSUE TITLE
Implement AttentiveFP using GRU

### ISSUE THREAD

**Issue by thegodone** [rusty1s/pytorch_geometric#2147]
From now on, we recommend using our discussion forum (https://github.com/rusty1s/pytorch_geometric/discussions) for proposing new features.

## 🚀 Feature

Can you implement this graph model : AttentiveFP which looks attractive in practice see papers

## Motivation

Two papers including a benchmarks shows great results with this code
https://pubs.acs.org/doi/10.1021/acs.jmedchem.9b00959
https://jcheminf.biomedcentral.com/articles/10.1186/s13321-020-00479-8

## Additional context
I found two implementation the original github https://github.com/OpenDrugAI/AttentiveFP/blob/master/code/AttentiveFP/AttentiveLayers.py
and this dgl port 
https://github.com/awslabs/dgl-lifesci/blob/master/python/dgllife/model/model_zoo/attentivefp_predictor.py

dgl-lifesci claims to have a x5 faster code compair to original source code.

I hope we can have a better speed with PyG

**Comment by rusty1s** [comment]
Sure, we can try to add it. Are you interested in contributing an inititial version?

**Comment by thegodone** [comment]
I can have try sure

**Comment by thegodone** [comment]
import torch
from torch import Tensor
from torch.nn import Linear, BatchNorm1d, Dropout
from torch.nn import Parameter as Param
import torch.nn.functional as F
from torch_geometric.nn import global_add_pool
from torch_sparse import matmul
from torch_geometric.nn.conv import MessagePassing
from torch_geometric.utils import softmax
from torch_geometric.typing import PairTensor, Adj, OptTensor, Size

from typing import Union


class AttentionAtomEmbedding(MessagePassing):
    """
    This function does only the atom embedding, not the molecule embedding
    """

    def __init__(self, atom_in_channels: int, bond_in_channels: int,  fingerprint_dim: int, dropout: float, bias: bool = True, **kwargs):
        super(AttentionAtomEmbedding, self).__init__()

        self.atom_in_channels = atom_in_channels
        self.bond_in_channels = bond_in_channels
        self.fingerprint_dim = fingerprint_dim
        
        # central atom feature only
        self.lin1 = Linear(atom_in_channels, fingerprint_dim, bias=bias)
        # feature atom & bond
        self.lin2 = Linear(atom_in_channels + bond_in_channels, fingerprint_dim, bias=bias)
        # align
        self.lin3 = Linear(2*fingerprint_dim,1, bias=bias)
        # attend = W need to find the correct dimensions
        self.W = torch.nn.Parameter(torch.rand(3,4,5))
        self.W.requires_grad = True 
        
        self.rnn =  torch.nn.GRUCell(fingerprint_dim, fingerprint_dim)
        self.dropout = Dropout(p=dropout)

    def forward(self, x: Union[Tensor,PairTensor], edge_index: Adj,
                edge_attr: OptTensor = None, size: Size=None) -> Tensor:
        
        out = self.propagate(edge_index, x = x, edge_attr=edge_attr, size=size)
        return out
    
    def message(self, x_i, x_j, edge_attr: OptTensor, size) -> Tensor:
        # use leakyrelu instead of relu
        # atom_feature => lin1
        h_v = F.leaky_relu(self.lin1(x_i))
        
        # neighbor_feature => lin2
        print(x_j.shape)
        print(edge_attr.shape)
        
        n_u = torch.cat([x_j, edge_attr], dim=-1)
        print(n_u.shape)
        h_u = F.leaky_relu(self.lin2(z))
        # feature_align
        h   = torch.cat([h_v, h_u], dim=-1)
        # align score
        e_vu = F.leaky_relu(self.lin3(h))
        # attention_weight
        a_vu = softmax(e_vu, edge_index, size)
        # context ouput: AxWxH
        tmp = torch.mm(torch.mm(a_vu,self.W), self.dropout(h_v))
        C_v = F.elu(scatter_add(tmp, edge_index, dim=0))
        # atom_feature output
        h_v = self.rnn(C_v, h_v)
        return h_v


class AttentiveMolEmbedding(torch.nn.Module):
    """
    This function does the molecule embedding
    """
    def __init__(self, radius:int, T: int, out_channels: int):
        super(AttentiveMolEmbedding, self).__init__()
        # need to find the correct dimensions 
        self.W = torch.nn.Parameter(torch.rand(3,4,5))
        self.W.requires_grad = True
        self.lin1 = Linear(1,1)
        # let start with one AAE with 49 atom features and 6 bond features and FP of 200
        self.atom_embedding =  AttentionAtomEmbedding(atom_in_channels = 49, bond_in_channels = 6,  fingerprint_dim = 200, dropout=0.3)
        self.rnn = torch.nn.GRUCell(out_channels, out_channels)
        self.lin_final = Linear(1,1)
        
    def forward(self, data):
        x, edge_index, batch, edge_attr = data.x, data.edge_index, data.batch, data.edge_feat
        
        h_s = global_add_pool(x, batch)
        # only one loop there:
        h_v =  self.atom_embedding(x, edge_index, edge_attr)
        
        e_sv = F.leaky_relu(self.lin1(torch.cat([h_s, h_v], dim=-1)))
        a_sv = softmax(e_sv, edge_index, batch)
        tmp = torch.mm(torch.mm(a_sv,self.W), out)
        C_s = F.elu(scatter_add(tmp, edge_index, dim=0))
        h_s = self.rnn(C_s, h_s)
        out  = self.lin_final(h_s)
        return out
        
                         
initial code with only one loop not all dimensions set but I have an issue to continue, can you help me ?

![image](https://user-images.githubusercontent.com/1186658/109656345-9d3e8b80-7b64-11eb-85db-8ac3b365a06a.png)

**Comment by rusty1s** [comment]
It looks like `edge_feat.size(0)` is different from `edge_index.size(1)`, which is not allowed. In general, each `edge_feat[i]` should correspond to the edge denoted in `edge_index[:, i]`.

**Comment by thegodone** [comment]
Good I fix this edge_attribute issue now I have the following issue:

![image](https://user-images.githubusercontent.com/1186658/110152237-c455c100-7de1-11eb-99cc-f0944b2095fe.png)
![image](https://user-images.githubusercontent.com/1186658/110152305-d7689100-7de1-11eb-8fdd-7548bfad080b.png)
![image](https://user-images.githubusercontent.com/1186658/110152364-ed765180-7de1-11eb-86b0-53dead3eb892.png)

thanks for your help

here the original code:
![image](https://user-images.githubusercontent.com/1186658/110152635-447c2680-7de2-11eb-80d2-cc23d0a56aa5.png)

**Comment by thegodone** [comment]
Hello I come to an end in the way PyG works:

in the original code, they need to propagate using "Node and Edge" but in PyG. I see only Node propagation so the edge signal is lost in the message passing propagation. 

![image](https://user-images.githubusercontent.com/1186658/110321411-ec2b6b80-8011-11eb-93a5-4dc9c94c34d4.png)

Do you see/ have a way to handle this issue ?

I've attached a self executable notebook. I stick to the original code structure as much as I can.

I only make one loop for the moment just to select the accurate functions:

[AttFP-PyG.zip](https://github.com/rusty1s/pytorch_geometric/files/6101271/AttFP-PyG.zip)

I copy the logs of the original code too 
[original pytorch logs.txt](https://github.com/rusty1s/pytorch_geometric/files/6101289/original.pytorch.logs.txt)

thanks for help.

Guillaume

**Comment by thegodone** [comment]
Ok now I check your runtime branch where you compare the dgl and PyG GATconv speed. I see that is maybe a better start to implement the AttFP ?

**Comment by rusty1s** [comment]
It looks like you want to concatenate features of different size:
```python
mol_cat_feature = torch.cat([mol_feature, activated_features], dim=-1)
```
with
```
mol_feature torch.Size([1, 200])
activated_features torch.Size([7, 200])
```
PyTorch does not automatically broadcast these tensors in `torch.cat`, so you do have to do this by yourself:
```python
mol_feature_expanded = model_feature.expand(activated_features.size(0), -1)
mol_cat_feature = torch.cat([mol_feature_expanded, activated_features], dim=-1)
```

**Comment by thegodone** [comment]
I see thanks for this one. I see also another issue the message / propagation is on node and we need to switch to edge message / propagation can I use your code in runtime branch for that ?

**Comment by rusty1s** [comment]
I do not recommend using other branches as they are pretty old and might not function well. I'm not sure what you mean with edge propagation, as your pseudo-code does not perform any edge propagation. The actual atom embedding implementation is really similar to the one we use in `GATConv`, which can be fully implemented in `message()`. After aggregation, one then applies the `GRU` layer on top of it.

**Comment by thegodone** [comment]
AttFP-DGL looks not reproducing results compair to original code. I found a PyG molecule by molecule code. It's in line to the initial architecture I provide already. But the PyG per molecule is very slow so I need to check how to combine the two codes.

**Comment by rusty1s** [comment]
Happy to integrate and benchmark AttFP once you have a running version :) Feel free to contribute.

**Comment by thegodone** [comment]
I test the current PyG one by one molecule code I found and it's clearly not matching the accuracy of the original code so. I stop on this one. I continue with the current version I made.

**Comment by thegodone** [comment]
I add my initial code here : https://github.com/thegodone/AttFP-PyG

let me know if you can improve speed.

I got the same result as in the paper on ESOL target, but it's slower than the original code without PyG

I think you know better than me all the tips in PyG...

![image](https://user-images.githubusercontent.com/1186658/110912017-bd640c80-8313-11eb-9f96-f19ff37b9780.png)

**Comment by thegodone** [comment]
I think I got solution running for T = R = 1

[AttFP2.zip](https://github.com/rusty1s/pytorch_geometric/files/6130558/AttFP2.zip)

But I don't see how to propagate hv1 to the neighbours see the image to get the node embedding ?

![image](https://user-images.githubusercontent.com/1186658/110952205-3c723880-8346-11eb-9d78-a25ef7019b3d.png)

Any idea ?

**Comment by rusty1s** [comment]
This looks similar to utilizing a `GATConv`, e.g.:
```python
h_0 = ...
C_0 = gat_conv(h_0, edge_index)
h_1 = gru(h_0, C_0)
```

**Comment by thegodone** [comment]
I think a have made a cleaner code now. 
last question, I don't see how to propagate for Radius = 2, ie how to get a mask to selection nodes connected at Radius 2 of a given node during the AtomEmbedding process. 
Do you have an idea ?

**Comment by rusty1s** [comment]
Does this operator really do that? I doubt it. In general, stacking GNN layers will result in aggregating a constantly larger range of neighbors. I think that all you have to do is stacking those layers.

**Comment by thegodone** [comment]
I think yes. They create sort of Morgan / ECFP atom aggregation at each AtomEmbedding of Radius 1, 2 with a mask (look at the original code https://github.com/OpenDrugAI/AttentiveFP/blob/master/code/AttentiveFP/AttentiveLayers.py). In DGL they did not make this trick and the models are not as good as in the original code. 
my last code following there supplementary paper pseudo code:
[attFP-PyG.py.zip](https://github.com/rusty1s/pytorch_geometric/files/6151087/attFP-PyG.py.zip)
